File: Trees/Python/Lowest Common Ancestor of a Binary Tree.py
# Problem link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/description/
# Time: O(n)
# Space: O(n) for call stack
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.dfs(root, p, q)
    
    def dfs(self, root, p, q):
        if not root or root == p or root == q:
            return root
        
        left = self.dfs(root.left, p, q)
        right = self.dfs(root.right, p, q)

        if not left:
            return right
        elif not right:
            return left
        return root

File: Trees/Java/Lowest Common Ancestor of a Binary Tree.java
// Problem link: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/description/
// Time: O(n)
// Space: O(n) for call stack

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        return this.dfs(root, p, q);
    }

    private TreeNode dfs(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root == p || root == q) {
            return root;
        }

        TreeNode left = this.dfs(root.left, p, q);
        TreeNode right = this.dfs(root.right, p, q);

        if(left == null) {
            return right;
        }
        else if(right == null) {
            return left;
        }
        return root;
    }
}
